collector/os_release.go

```diff
@@ -0,0 +1,174 @@
+	"regexp"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	envparse "github.com/hashicorp/go-envparse"
+	"github.com/prometheus/client_golang/prometheus"
+	kingpin "gopkg.in/alecthomas/kingpin.v2"
+)
+
+var (
+	collectorOsPrefix = kingpin.Flag("collector.os.prefix", "path prefix to use for os collector metrics for testing").Default("").String()
+	etcOsRelease      = "/etc/os-release"
+	usrLibOsRelease   = "/usr/lib/os-release"
+	versionRegex      = regexp.MustCompile(`^[0-9]+\.?[0-9]*`)
+)
+
+type osRelease struct {
+	Name            string
+	ID              string
+	IDLike          string
+	PrettyName      string
+	Variant         string
+	VariantID       string
+	Version         string
+	VersionID       string
+	VersionCodename string
+	BuildID         string
+	ImageID         string
+	ImageVersion    string
+}
+
+type osReleaseCollector struct {
+	infoDesc             *prometheus.Desc
+	logger               log.Logger
+	mtimeEtcOsRelease    time.Time
+	mtimeUsrLibOsRelease time.Time
+	os                   *osRelease
+	version              float64
+	versionDesc          *prometheus.Desc
+}
+
+func init() {
+	registerCollector("os", defaultEnabled, NewOSCollector)
+}
+
+// NewOSCollector returns a new Collector exposing os-release information.
+func NewOSCollector(logger log.Logger) (Collector, error) {
+	return &osReleaseCollector{
+		logger: logger,
+		infoDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "os", "info"),
+			"A metric with a constant '1' value labeled by build_id, id, id_like, image_id, image_version, "+
+				"name, pretty_name, variant, variant_id, version, version_codename, version_id.",
+			[]string{"build_id", "id", "id_like", "image_id", "image_version", "name", "pretty_name",
+				"variant", "variant_id", "version", "version_codename", "version_id"}, nil,
+		),
+		versionDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "os", "version"),
+			"Metric containing the major.minor part of the OS version.",
+			nil, nil,
+		),
+	}, nil
+}
+
+func parseOsRelease(r io.Reader) (*osRelease, error) {
+	env, err := envparse.Parse(r)
+	return &osRelease{
+		Name:            env["NAME"],
+		ID:              env["ID"],
+		IDLike:          env["ID_LIKE"],
+		PrettyName:      env["PRETTY_NAME"],
+		Variant:         env["VARIANT"],
+		VariantID:       env["VARIANT_ID"],
+		Version:         env["VERSION"],
+		VersionID:       env["VERSION_ID"],
+		VersionCodename: env["VERSION_CODENAME"],
+		BuildID:         env["BUILD_ID"],
+		ImageID:         env["IMAGE_ID"],
+		ImageVersion:    env["IMAGE_VERSION"],
+	}, err
+}
+
+func (c *osReleaseCollector) updateIfChanged(f *os.File, mtime *time.Time) error {
+	stat, err := f.Stat()
+	if err != nil {
+		return err
+	}
+
+	t := stat.ModTime()
+	if t != *mtime {
+		level.Debug(c.logger).Log("msg", "file modification time has changed",
+			"file", f.Name(), "old_value", mtime, "new_value", t)
+		*mtime = t
+		c.os, err = parseOsRelease(f)
```

Comment:
This is not thread safe. Updates to the collector struct need to be locked.

Reply:
Added an updateMutex to cover the struct updates.